test/e2e/network/traffic_distribution.go

```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
 		}
 	}
 
+	// Data structures for representing client and endpoint pods
+	type endpoint struct {
+		node *v1.Node
+		pod  *v1.Pod
+	}
+
+	type client struct {
+		node      *v1.Node
+		endpoints []*endpoint
+		pod       *v1.Pod
+	}
+
 	////////////////////////////////////////////////////////////////////////////
 	// Main test specifications.
 	////////////////////////////////////////////////////////////////////////////
 
-	ginkgo.When("Service has trafficDistribution=PreferClose", func() {
-		ginkgo.It("should route traffic to an endpoint that is close to the client", func(ctx context.Context) {
+	// doTrafficDistributionTest runs a test of a service with the given trafficDist,
+	// clients, and endpoints, ensuring that connections go to the expected endpoints.
+	doTrafficDistributionTest := func(ctx context.Context, trafficDist string, clients []*client, endpoints []*endpoint) {
+		nodesHaveZones := false
+		var servingPods []*v1.Pod
+		servingPodLabels := map[string]string{"app": f.UniqueName}
+		for i, ep := range endpoints {
+			node := ep.node.Name
+			zone := ep.node.Labels[v1.LabelTopologyZone]
+			if zone != "" {
+				nodesHaveZones = true
+			}
+			pod := e2epod.NewAgnhostPod(f.Namespace.Name, fmt.Sprintf("endpoint-%d-%s", i, node), nil, nil, nil, "serve-hostname")
+			ginkgo.By(fmt.Sprintf("creating a server pod %q on node %q in zone %q", pod.Name, node, zone))
```

Comment:
this is what I think we should call this `servers`, endpoint is the representation but this is the realization